airbyte-integrations/connectors/source-mongodb/lib/mongodb_stream.rb

```diff
@@ -0,0 +1,85 @@
+require_relative './airbyte_protocol.rb'
+
+class MongodbStream
+  PROPERTIES_DISCOVERY_LIMIT = 10_000
+  TYPE_ANY = 'ANY_TYPE'
+  TYPE_GROUPS = {
+    simple: 'SIMPLE_TYPE_GROUP',
+    complex: 'COMPLEX_TYPE_GROUP',
+  }
+
+  AIRBYTE_TYPES = {
+    boolean: 'boolean',
+    number: 'number',
+    integer: 'integer',
+    string: 'string',
+    object: 'object',
+    array: 'array',
+  }
+
+  TYPES_MAPPING = {
+    Float => AIRBYTE_TYPES[:number],
+    Integer => AIRBYTE_TYPES[:integer],
+    String => AIRBYTE_TYPES[:string],
+    DateTime => AIRBYTE_TYPES[:string],
+    TrueClass => AIRBYTE_TYPES[:boolean],
+    FalseClass => AIRBYTE_TYPES[:boolean],
+    Array => AIRBYTE_TYPES[:array],
+    Hash => AIRBYTE_TYPES[:hash],
+  }
+  FALLBACK_TYPE = AIRBYTE_TYPES[:string]
+
+  def initialize(collection:)
+    @collection = collection
+    @property_types = {}
+    @properties = {}
+  end
+
+  def discover
+    discover_property_types
+    discover_properties
+
+    AirbyteStream.from_dynamic!({
+      "name" => @collection.name,
+      "supported_sync_modes" => [SyncMode::FullRefresh],
+      "json_schema" => {
+        "properties": @properties
+      }
+    }).to_dynamic
+  end
+
+  private
+
+
+  def discover_property_types
+    @collection.find.limit(PROPERTIES_DISCOVERY_LIMIT).each do |item|
+      item.each_pair do |key, value|
+        @property_types[key] ||= Set[]
+        @property_types[key].add value.class
+      end
+    end
+  end
+
+  def discover_properties
+    @property_types.each_pair do |key, types|
+      airbyte_types = types.map do |type|
+        # Skip nil classes. No impact on actual column type
+        if type != NilClass
+          map_type(type)
+        end
+      end.compact
```

Comment:
is there a chance of getting nil in this `map`? I'm a little scared seeing the `compact` because it tells me that some columns might have a nil type and we are silently sweeping it under the rug.

Reply:
Yes, there is. MongoDB does not have anything like table definition, thus we have to define column types from actual attribute values. When we discover schema, we use only the first 10k records in each collection. There is a chance that we encounter 10k `null` values and can not define column type properly. In that case we fallback to the `string` type and rely on custom data transformation of imported data.

Reply:
Finally, I decided to go a different way. I reworked the `discover` command to run the map-reduce command which should be sufficient even for large clusters and will return all unique document props in the collection. Then trying to determine prop type selecting 10k records from the collection where this prop is not empty.

I believe that approach will work for all cases. Nevertheless, I will update the readme